8292660: C2: blocks made unreachable by NeverBranch-to-Goto conversion are removed incorrectly

robcasloz opened the pull request:
This changeset addresses three issues in the current removal of unreachable blocks after NeverBranch-to-goto conversion (introduced recently by [JDK-8292285](https://bugs.openjdk.org/browse/JDK-8292285)):

1. The [unreachable block removal and pre-order index update loop](https://github.com/openjdk/jdk/blob/7b5f9edb59ef763acca80724ca37f3624d720d06/src/hotspot/share/opto/block.cpp#L613-L621) skips the block next to the removed one, and iterates beyond the end of the block list (`PhaseCFG::_blocks`). Skipping blocks can lead to duplicate pre-order indices (`Block::_pre_order`) and/or pre-order indices greater than the size of the block list, causing problems in later transformations.

2. The [outer block traversal loop](https://github.com/openjdk/jdk/blob/7b5f9edb59ef763acca80724ca37f3624d720d06/src/hotspot/share/opto/block.cpp#L698-L729) iterates beyond the end of the block list whenever one or more unreachable blocks are removed.

3. Transitively unreachable blocks (such as B10 in the following example), arising in methods with multiple infinite loops, are not removed:

![transitive](https://user-images.githubusercontent.com/8792647/186109043-416213b7-8735-41de-9910-acf0997db095.png)

This changeset addresses issues 2 and 3 by decoupling NeverBranch-to-goto conversion from removal of unreachable code. Instead of removing the blocks eagerly, the removal is postponed to a later phase that works in an iterative worklist fashion, making it possible to remove transitively unreachable blocks such as B10 in the above example. Postponing removal to a later phase (where `get_block(i)->_pre_order == i` holds) also simplifies addressing issue 1: in the changeset, it is sufficient to iterate over the blocks that follow the removed block in the block list to decrement their `_pre_order` index.

#### Testing

- tier1-3 (windows-x64, linux-x64, linux-aarch64, and macosx-x64; release and debug mode).
- tier4-7 (linux-x64; debug mode).
- fuzzing (~1 h. on each platform).

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292660](https://bugs.openjdk.org/browse/JDK-8292660): C2: blocks made unreachable by NeverBranch-to-Goto conversion are removed incorrectly


### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [7a95e6fc](https://git.openjdk.org/jdk/pull/9976/files/7a95e6fcd3979fb1ae5ea6e28de0bca283363679)
 * [Roland Westrelin](https://openjdk.org/census#roland) (@rwestrel - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9976/head:pull/9976` \
`$ git checkout pull/9976`

Update a local copy of the PR: \
`$ git checkout pull/9976` \
`$ git pull https://git.openjdk.org/jdk pull/9976/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9976`

View PR using the GUI difftool: \
`$ git pr show -t 9976`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9976.diff">https://git.openjdk.org/jdk/pull/9976.diff</a>

</details>

vnkozlov commented on src/hotspot/share/opto/block.cpp:
```diff
@@ -738,6 +730,41 @@ void PhaseCFG::remove_empty_blocks() {
+      if (block == dead) {
+        _blocks.remove(i);
+        _number_of_blocks--;
+        i--; // Ensure that we visit the block following the removed one.
```
May be `continue;` here to skip following checks for `dead` block.

vnkozlov commented on src/hotspot/share/opto/block.cpp:
```diff
@@ -738,6 +730,41 @@ void PhaseCFG::remove_empty_blocks() {
+        _number_of_blocks--;
+        i--; // Ensure that we visit the block following the removed one.
+      }
+      if (block->_pre_order > dead->_pre_order) {
+        // Enforce contiguous pre-order indices (assumed by PhaseBlockLayout).
+        block->_pre_order--;
+      }
+      if (block->_rpo > dead->_rpo) {
+        // Enforce contiguous reverse post-order indices as well, for sanity.
+        block->_rpo--;
+      }
+    }
```
Are you sure it will work for complex case? May be recalculate RPO and order after you removed dead blocks from graph.

robcasloz commented:
Thanks for the review, Vladimir! I have updated the changeset to postpone removal of unreachable blocks to after `PhaseCFG::fixup_flow()`. This makes the handling of block indices simpler and less error-prone: at this stage `Block::_pre_order` does not contain pre-order indices but simple block list indices, and `Block::_rpo` does not contain valid reverse post-order indices anymore (they are invalidated by block insertions in `PhaseCFG::insert_goto_at()`). I have also added a `continue` as suggested.

vnkozlov reviewed (approved):
Looks good.

robcasloz commented:
> Looks good.

Thanks for re-reviewing!

RealFYang commented:
Verified that this also resolves issue: https://bugs.openjdk.org/browse/JDK-8292859

robcasloz commented:
> Verified that this also resolves issue: https://bugs.openjdk.org/browse/JDK-8292859

Thanks for checking, Fei.

rwestrel commented on src/hotspot/share/opto/block.cpp:
```diff
@@ -954,6 +945,52 @@ void PhaseCFG::fixup_flow() {
+  // Now remove all blocks that are transitively unreachable.
+  while (unreachable.size() > 0) {
+    Block* dead = unreachable.pop();
+    for (uint i = 0; i < _number_of_blocks; i++) {
```
Wouldn't we want to iterate backward here and adjust block->_pre_order until we hit dead?

robcasloz replied:
Thanks for the suggestion, Roland! Yes, since `get_block(i)->_pre_order == i` holds at this stage, it is enough with iterating over the elements that succeed `dead` in the block list. I have updated the changeset accordingly.

rwestrel reviewed (approved):
Looks good to me.

robcasloz commented:
> Looks good to me.

Thanks for reviewing, Roland!

robcasloz commented:
/integrate
